### PR TITLE
Add simple TUF role metadata model

### PR DESCRIPTION
**Pull request by lukpueh**
This PR copies over the new `api/metadata.py` module (+ tests) added in #1060, which has shifted focus to implementing a Hashicorp Vault interface.

Co-authorship by @trishankatdatadog, @joshuagl, @sechkova and me plus note-worthy commit message details are preserved in a single squashed commit.

Missing TODOs items for this PR include at least.

- [x] add code documentation (docstrings)
- [x] finalize signature creation and verification interface
- [ ] add input validatation

See doc header of `api/metadata.py` for more details

### CODE REVIEW

File: tests/test_api.py
Comment (trishankatdatadog) [inline tests/test_api.py]:
```suggestion
""" Unit tests for api/metadata.py
```

**Comment by sechkova** [comment]
_I will be leaving some thoughts here while looking at the way the client side (Updater) may use the api/metadata._

- There are several places in the code where Metadata needs to be loaded by passing a file object instead of a filename, in particular when Updater downloads temporary metadata files and verifies them. Should we plan for adding another method like `read_from_json_object` or we may use the Metadata() constructor directly by passing the file object already serialised as a dictionary and not complicating this api?

**Comment by lukpueh** [comment]
> * There are several places in the code where Metadata needs to be loaded by passing a file object instead of a filename, in particular when Updater downloads temporary metadata files and verifies them. Should we plan for adding another method like `read_from_json_object` or we may use the Metadata() constructor directly by passing the file object already serialised as a dictionary and not complicating this api?

Great idea! What about the following interface?

```python
# Classmethod on all classes
def from_dict(json_dict):
    """
    Initialize object from Python dictionary, and call from_dict for any
    inner complex attribute that is represented by a class.

    In the case of Metadata.from_dict(), it must dispatch to the corresponding
    from_dict of the base class of Signed, identified by
    json_dict['signed']['_type'].

    NOTE: This is basically a mix of the constructor and load_from_json and
    the inverse of `as_dict` in the current version of metadata.py. """

# Classmethods on Metadata (and optionally on Signed)
def from_json(json_object):
    """Deserialize a json string and pass to from_dict(). """

def from_json_file(filename):
    """Read file from storage and pass to from_json(). """
```

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]:
```suggestion
                to the list of signatures or replace any existing signatures.
                The default behaviour is to replace signatures.
```

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]:
```suggestion
            and/or if more than one signatures are found for the keyid.
```

**Comment by lukpueh** [comment]
Hey @sechkova, I added an `Metadata.from_json(metadat_json_string)` in a7a33ce. Let me know if this is what you had in mind.  1cdae24 generally cleans up the deserialization interface as I proposed in https://github.com/theupdateframework/tuf/pull/1112#issuecomment-683701445 (see detailed commit message).

@woodruffw, @joshuagl, @trishankatdatadog, I'd say this is ready for review aka. throwing rocks at it. It still misses some features  but what's there is tested, documented and should serve our purposes for PEP 458. I think we can work off the todo list from `metadata.py`'s doc header independently in follow-up PRs.

File: tuf/api/metadata.py
Comment (lukpueh) [inline tuf/api/metadata.py]: Now a `datetime`object.

File: tuf/api/metadata.py
Comment (lukpueh) [inline tuf/api/metadata.py]:
```suggestion
    def from_dict(cls, signed_dict: JsonDict) -> 'Signed':
```

File: tuf/api/metadata.py
Comment (lukpueh) [inline tuf/api/metadata.py]: Btw. what do others think about the [ValidationMixin](https://github.com/in-toto/in-toto/blob/74da7aabce663a5539ad3e1e17b2c871eeaa227a/in_toto/models/common.py#L27-L40) methods we use in in-toto? (also see [`Layout._validate_*`](https://github.com/in-toto/in-toto/blob/74da7aabce663a5539ad3e1e17b2c871eeaa227a/in_toto/models/layout.py#L420-L438) methods for example usage) 

I think it's a commendable way of doing this (credits to @SantiagoTorres)

**Review by trishankatdatadog — APPROVED** [review]
Great job, @lukpueh! 🎉 

I don't have strong opinions about this PR: it looks largely like the Metadata API I originally proposed, except for changes like removing KeyRing and Threshold so on (we can argue about that later). As long as @woodruffw and @joshuagl  also approve, I'm 👍🏽

File: tests/test_api.py
Comment (joshuagl) [inline tests/test_api.py:57]: 8 space indentation on a continuation?

File: tuf/api/metadata.py
Comment (joshuagl) [reply tuf/api/metadata.py]: That `ValidationMixin` looks nice! I'd be happy to see it used here in tuf

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Do we still need an iso8601 dependency if we're only targeting Python 3.5+, can we just use `datetime.fromisoformat()` here?

File: tests/test_api.py
Comment (joshuagl) [inline tests/test_api.py]: I presume the goal would be to revert this patch once all Python 2.7 supporting code has been removed?

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]:
```suggestion
    Objects with base class Signed are usually included in a Metadata object
```

File: tuf/api/metadata.py
Comment (joshuagl) [reply tuf/api/metadata.py]: I prefer the alternative approach of raising an exception if no signature is found, or more than one signature is found, for the keyid.

**Review by joshuagl — COMMENTED** [review]
Very nice streamlining of the metadata API PR. Thanks @lukpueh! I have a few minor questions and comments inline. I'm also curious to understand why `version` and `spec_version` have different types? `version: int` and `spec_version: str`, is there any reason `spec_version` can't be an int?

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: I think I do too. Any advice on the exception type? At some point we should discuss [exception taxonomy guidelines](https://github.com/theupdateframework/tuf/projects/1#card-44476165). But we shouldn't block on that.

Comment (joshuagl) [reply tuf/api/metadata.py]: We don't seem to have an exception in `tuf.exceptions` or `securesystemslib.exceptions` that would fit, we could either throw a generic exception _or_ define a new one. Either way, it will probably need some cleanup once guidelines are defined.

File: tests/test_api.py
Comment (lukpueh) [reply tests/test_api.py:57]: Hm. Good question. I always thought [PEP 8 required double-indentation for hanging indent](https://www.python.org/dev/peps/pep-0008/#indentation), but it is only a suggestion, and also only to distinguish it from a subsequent nested block. 

IMO it's a good idea to always differentiate line continuation indentation from functional indentation, even if there is no subsequent nested block. But I'm fine either way.

FYI, the [google style guide does not say anything about double indent](https://google.github.io/styleguide/pyguide.html#indentation) but has some [examples that could be interpreted that way](https://google.github.io/styleguide/pyguide.html#s3.2-line-length).

Comment (joshuagl) [reply tests/test_api.py:57]: > IMO it's a good idea to always differentiate line continuation indentation from functional indentation, even if there is no subsequent nested block. But I'm fine either way.
> 
That seems like a reasonable coding style for us to adopt. We should document it and teach the linter about it 😃

**Comment by lukpueh** [comment]
> version: int and spec_version: str, is there any reason spec_version can't be an int?

Yes, `spec_version` is in semver format.

**Comment by lukpueh** [comment]
Thanks for the reviews, @joshuagl and @trishankatdatadog!. I addressed the minor comments in 0a9fc93, which I'm happy to squash with previous commits before merging. The more serious change of `Metadata.verify` behaviour is in a separate commit, to make review easier. :)

File: tests/test_api.py
Comment (joshuagl) [inline tests/test_api.py:20]:
```suggestion
# TODO: Remove case handling when fully dropping support for versions < 3.6
```

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: 🎉

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py:299]: Awesome, thanks for adding this!

**Review by joshuagl — APPROVED** [review]
Thanks @lukpueh, this looks solid. I'm very happy to see this long-discussed model implemented.

One minor typo in a comment, but otherwise this looks good to rebase and merge.

File: tuf/api/metadata.py
Comment (joshuagl) [reply tuf/api/metadata.py:299]: We have three uses of `iso8601.parse_date` in the current codebase, so this might be a good candidate for a helper function in future.

File: tests/test_api.py
Comment (lukpueh) [reply tests/test_api.py:20]: 🤦 thanks for catching this!

**Comment by lukpueh** [comment]
Thanks again for the reviews! I just cleaned up the git history, fixed a minor inconsistency and removed the todo list after created GitHub issues for each item.
The diff of the new force-pushed tip to your last review can be seen here: https://github.com/lukpueh/tuf/compare/simple-tuf-api-before-rebase..simple-tuf-api

**Pull request merged**
